### PR TITLE
WorkbookRepository

### PR DESCRIPTION
**Pull request by aunger**
Implement WorkbookRepository to build new data model objects.
Co-requires https://github.com/WycliffeAssociates/otter-jvm/pull/441

### CODE REVIEW

File: src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/Take.kt
Comment (aunger) [inline src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/Take.kt:17]: @sarabiaj The [RX Maybe](http://reactivex.io/RxJava/javadoc/io/reactivex/Maybe.html) isn't applicable here. We might have talked about [Arrow's Option](https://arrow-kt.io/docs/arrow/core/option/) or some implementation of a Haskell-style [Maybe](https://gist.github.com/ramseyboy/91bcc85cb5f28e47c9b5464c1501a983).  Those are pretty awesome, but their APIs for converting to/from nulls are more verbose than just using a data class.  I guess their opinion is that you should only rarely be using or converting nullables. That's sensible, but not the path we're on.

File: src/main/kotlin/org/wycliffeassociates/otter/common/domain/resourcecontainer/project/markdown/MarkdownProjectReader.kt
Comment (KJoslyn) [inline src/main/kotlin/org/wycliffeassociates/otter/common/domain/resourcecontainer/project/markdown/MarkdownProjectReader.kt:91]: This is probably stating the obvious, but I suppose we could stop using the label field to distinguish titles from bodies now. I don't remember what else the label field is used for

**Review by KJoslyn — APPROVED** [review]

File: src/main/kotlin/org/wycliffeassociates/otter/common/persistence/repositories/ITakeRepository.kt
Comment (jsarabia) [inline src/main/kotlin/org/wycliffeassociates/otter/common/persistence/repositories/ITakeRepository.kt]: If the name "obj" is trying to adhere to existing practices, know that I never meant to let "obj" get through PR's.

Let's stick with "take" for the variable name

File: src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/HasChildBookElements.kt
Comment (jsarabia) [inline src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/HasChildBookElements.kt]: I'm uneasy about the name of this interface (but also ResourceContainerInfo), as it is unclear what it does and what its use is for. 

I roughly know the problem it's intending to solve, but am a little fuzzy just due to how much time has passed- which helps indicate that it doesn't really hold up to being self documenting.

My thought process is mostly as follows: the "has" prefix implies a boolean state: having something or not having something. That something is described by "ChildBookElements," so my expectation is there should be a method I can use that tells me whether or not the object carrying this interface _has_ child book elements. I do not expect the interface to contain such elements, especially if it does not explicitly answer the question the name poses (I suspect the way to check if the object that has this interface actually has child book elements is to do a .isEmpty() call on the list, or to subscribe on children (which now makes my boolean check async (uncomfortable)), but this kinda feels like the interface isn't what is actually providing me the functionality explicitly, despite what the name suggests).

I'm going to tag @natibekele and @KJoslyn here because hopefully me breaking this down is useful for understanding what I value in naming variables/methods/classes/etc and how I interpret them.


Is it intended that this would contain anything other than "helps"? (ie tn, tq?)? If no, the scope can be restricted to just the helps, and doesn't need to be an abstract concept (BookElement).

File: src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/ResourceContainerInfo.kt
Comment (jsarabia) [inline src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/ResourceContainerInfo.kt]: ResourceInfo is almost a better name here in that ResourceContainerMetadata is so similar, and because this doesn't exactly have anything to do with ResourceContainers (it is referring to the slug and title of a resource).

However, perhaps this highlights a problem we've overlooked in the past that applies to more things? Title and slug are largely separate fields for use of hash lookups, sometimes for apis but often to separate internal app use and user facing text (ie, already localized). As a result virtually everything has a slug and a title field.

Perhaps the data class here is just... Title; which exists in the data model package?

Not sure how I feel about that, just my initial thoughts.

File: src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/HasChildBookElements.kt
Comment (jsarabia) [reply src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/HasChildBookElements.kt]: also @mXaln

Comment (aunger) [reply src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/HasChildBookElements.kt]: The "has" _interface_ naming pattern doesn't mean you should expect to find a "has" _method_ inside.  Implementation of the interface means the class "has" the thing in question and you should be able to get at that thing.
See examples:
- https://github.com/search?l=Kotlin&q=%22interface+has%2A%22&type=Code
- https://docs.microsoft.com/en-us/dotnet/api/?term=has
- http://old.haxe.org/manual/2_types#interfaces
- https://stackoverflow.com/questions/5266907/how-to-come-up-with-clearer-interface-names

I can certainly rename this interface to be `BookElementParent` or `BookElementContainer` if one of those is more clear.

File: src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/ResourceContainerInfo.kt
Comment (aunger) [reply src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/ResourceContainerInfo.kt]: `ResourceBook` was my initial concept and we went back and forth on that for a while until I landed on `ResourceContainerInfo`.  In my mind, this is really just a whittled down version of `ResourceContainerMetadata`, and has everything to do with resource containers.  In fact, I'd probably just replace this class with `ResourceContainerMetadata` if you were OK with that.

File: src/main/kotlin/org/wycliffeassociates/otter/common/domain/resourcecontainer/project/markdown/MarkdownProjectReader.kt
Comment (aunger) [reply src/main/kotlin/org/wycliffeassociates/otter/common/domain/resourcecontainer/project/markdown/MarkdownProjectReader.kt:91]: Yep! Forthcoming (in another PR).

File: src/main/kotlin/org/wycliffeassociates/otter/common/persistence/repositories/ITakeRepository.kt
Comment (aunger) [reply src/main/kotlin/org/wycliffeassociates/otter/common/persistence/repositories/ITakeRepository.kt]: I didn't choose it on purpose. I don't know whether it came from code pasted from elsewhere, or some IDEA auto-refactor artifact. I'll change it.

File: src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/HasChildBookElements.kt
Comment (aunger) [reply src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/HasChildBookElements.kt]: Oh, also see my new tab on draw.io; we might not be on the same page re design.

File: src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/ResourceContainerInfo.kt
Comment (jsarabia) [reply src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/ResourceContainerInfo.kt]: My hesitancy comes from the existence of this class:
https://github.com/WycliffeAssociates/otter-common/blob/dev/src/main/kotlin/org/wycliffeassociates/otter/common/data/model/ResourceMetadata.kt

Though admittedly I had thought it was named ResourceContainerMetadata.

File: src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/HasChildBookElements.kt
Comment (jsarabia) [reply src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/HasChildBookElements.kt]: I believe interfaces are typically meant to describe Is-A relationships rather than Has-A, similar to inheritance. I had not seen interfaces with "Has*" before, though it clearly has been done. At the same time, from looking through the github search results I don't really see many large well known projects or repos by companies, even when searching by Java- most are individuals. The Microsoft link is much more compelling, but at the same time I notice that it appears to mostly violate their own style guide: https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-classes-structs-and-interfaces

Of course, there is nothing _wrong_ with the code above, it just comes down to code style. I do think that BookElementContainer is much more clear regarding the intent of the interface as well as how it expresses a trait of the class implementing it.

File: src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/ResourceContainerInfo.kt
Comment (aunger) [reply src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/ResourceContainerInfo.kt]: Renamed as `ResourceInfo`.

File: src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/HasChildBookElements.kt
Comment (aunger) [reply src/main/kotlin/org/wycliffeassociates/otter/common/data/workbook/HasChildBookElements.kt]: Renamed as `BookElementContainer`.

File: src/main/kotlin/org/wycliffeassociates/otter/common/persistence/repositories/ITakeRepository.kt
Comment (aunger) [reply src/main/kotlin/org/wycliffeassociates/otter/common/persistence/repositories/ITakeRepository.kt]: Done.

**Pull request merged**
